crypto/dh/dh_group_params.c

```diff
@@ -141,11 +257,13 @@ int DH_get_nid(DH *dh)
                 dh->params.q = q;
                 q = NULL; /* set to NULL so it is not freed */
             }
-            dh->params.nid = sp_groups[i].nid; /* cache the nid */
-            dh->length = 2 * ifc_ffc_compute_security_bits(sp_groups[i].nbits);
-            dh->dirty_cnt++;
-            break;
         }
+        nid = dh->params.nid = dh_named_groups[i].uid; /* cache the nid */
+        dh->length =
+            2 * ifc_ffc_compute_security_bits(dh_named_groups[i].nbits);
+        dh->dirty_cnt++;
+        /* A matching p was found so break out of the loop */
+        break;
```

Comment:
The more I look at what's happening with `DH_get_nid()`, the more I cringe.  Why was `DH_get_nid()`, of all possible functions, selected as the one to have all these side effects?  Why isn't this caching functionality refactored to its own function? Why isn't this caching done when the rest of `dh->params` is generated?  Or loaded?

Reply:
I must be missing something.
The user can still set the params directly and potentially do one set at a time of any of p, q, &/or g in any order. So where do you cache the nid/q/length then? (And we then still end up with side effects on the sets - so either way you need to choose your poison).

Reply:
DH_get_nid() will be addressed in another PR after this one goes in.

Reply:
Is that a promise?  I'll hold you do it.